subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/compilerapi/constants/ConstantToDependentsMappingBuilder.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile.incremental.compilerapi.constants;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A builder helper class to construct the ConstantToDependentsMapping
+ */
+public class ConstantToDependentsMappingBuilder implements Serializable {
+
+    private final Set<String> visitedClasses;
+    private final List<String> dependents;
+    private final Map<Integer, Set<Integer>> privateDependentsIndexes;
+    private final Map<Integer, Set<Integer>> accessibleDependentsIndexes;
+    private final Map<String, Integer> classNameToIndex;
+
+    ConstantToDependentsMappingBuilder() {
+        this.visitedClasses = new HashSet<>();
+        this.dependents = new ArrayList<>();
+        this.classNameToIndex = new HashMap<>();
+        this.privateDependentsIndexes = new HashMap<>();
+        this.accessibleDependentsIndexes = new HashMap<>();
+    }
+
+    public ConstantToDependentsMappingBuilder addAccessibleDependents(int constantOriginHash, Collection<String> dependents) {
```

Comment:
May I ask why you're using the hashcode of the types here, instead of the fully qualified class name? If I'm not mistaken it means that 2 classes having the same hashcode would be considered the same thing, and therefore we would consider their constants to be from the same class.

I suppose this is an optimization to reduce the serialized size? If so, did you measure the impact?

Reply:
That's an idea yeah. And you are correct that can be a result. Luckily in that case we would just compile more classes and that is not an issue. But I figured there is an issue when we merge new and old mapping, since we also remove all removed classes. If we remove too much here, that is an issue.

After you pointed this out I did some quick tests with this version and version with string on some mid size projects with cca. 10k classes:
- there is no difference in the size of cache file previous-compilation.bin
- it's seems that there is no significant difference in gc time 
- incremental compile time in both versions was cca. same cca. 35s, I had filling that in some cases version with hashcode is slighly faster, maybe in some cases less than 10%, but I would have to measure it in some more controlled env. to actually prove that. It might be, that just my Mac was tired when I run a different version.

So overall I think due to correctness this should change to using fully qualified class name. Thanks for pointing that out.

If you want discuss anything you can also ping me on Slack if you want. I am not always there, but I check it from time to time during the day.

Reply:
Great, thanks. While you're at it, I made some more changes locally: replacing `Set<Integer>` with `IntSet` which is significantly more efficient:

```diff
diff --git a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/CompilationOutputsFixture.groovy b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/CompilationOutputsFixture.groovy
index 8da17666729..4cad21fb6bc 100644
--- a/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/CompilationOutputsFixture.groovy
+++ b/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/CompilationOutputsFixture.groovy
@@ -17,6 +17,7 @@
 package org.gradle.integtests.fixtures
 
 import groovy.io.FileType
+import org.codehaus.groovy.control.CompilerConfiguration
 import org.gradle.internal.FileUtils
 import org.gradle.util.internal.TextUtil
 
@@ -150,4 +151,14 @@ class CompilationOutputsFixture {
         }
         changed
     }
+
+    Object evaluate(@GroovyBuildScriptLanguage String script) {
+        CompilerConfiguration config = new CompilerConfiguration()
+        config.classpath.add("${targetDir}/$lang/$sourceSet".toString())
+        new GroovyShell(config).evaluate(script)
+    }
+
+    void execute(@GroovyBuildScriptLanguage String script) {
+        evaluate(script)
+    }
 }
diff --git a/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractCrossTaskIncrementalJavaCompilationIntegrationTest.groovy b/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractCrossTaskIncrementalJavaCompilationIntegrationTest.groovy
index 7636cee2ab0..2c67f11571b 100644
--- a/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractCrossTaskIncrementalJavaCompilationIntegrationTest.groovy
+++ b/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractCrossTaskIncrementalJavaCompilationIntegrationTest.groovy
@@ -191,6 +191,12 @@ abstract class AbstractCrossTaskIncrementalJavaCompilationIntegrationTest extend
                    "class Z { final static int x = Y.x;  }",
                    "class W {  }"]
         impl.snapshot { run language.compileTaskName }
+        impl.execute("""
+            assert X.x == 1
+            assert Y.x == 1
+            assert new Y().method() == 1
+            assert Z.x == 1
+        """)
 
         when:
         source api: ["class A { final static int x = 2; /* change */ void blah() { /* avoid flakiness by changing compiled file length*/ } }"]
@@ -198,6 +204,14 @@ abstract class AbstractCrossTaskIncrementalJavaCompilationIntegrationTest extend
 
         then:
         impl.recompiledClasses('X', 'Y')
+
+        and:
+        impl.execute("""
+            assert X.x == 2
+            assert Y.x == 1
+            assert new Y().method() == 2
+            assert Z.x == 1
+        """)
     }
 
     def "changing upstream constant causes compilation for downstream constants for binary expression"() {
diff --git a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/compilerapi/constants/ConstantToDependentsMapping.java b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/compilerapi/constants/ConstantToDependentsMapping.java
index 6198f250393..4171ac50ded 100644
--- a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/compilerapi/constants/ConstantToDependentsMapping.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/compilerapi/constants/ConstantToDependentsMapping.java
@@ -17,6 +17,7 @@
 package org.gradle.api.internal.tasks.compile.incremental.compilerapi.constants;
 
 import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
+import it.unimi.dsi.fastutil.ints.IntSet;
 import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
 import org.gradle.api.internal.cache.StringInterner;
 import org.gradle.internal.serialize.AbstractSerializer;
@@ -34,7 +35,8 @@
 
 /**
  * A data structure that holds constant to class mapping.
- * It consists of dependents class names `dependentClasses` and dependents `privateConstantDependents` and `accessibleConstantDependents` as `constant hash` => set(indexes) mapping.
+ * It consists of dependents class names `dependentClasses` and dependents `privateConstantDependents`
+ * and `accessibleConstantDependents` as `constant hash` => set(indexes) mapping.
  *
  * The idea here is that Strings take a lot of memory and there is one-to-many relationship between constant to dependents.
  * So that is why we store String names of dependents in a separate list and we access them via indexes.
@@ -48,13 +50,13 @@
     private transient final Set<String> visitedClasses;
 
     private final List<String> dependentClasses;
-    private final Map<Integer, Set<Integer>> privateConstantDependents;
-    private final Map<Integer, Set<Integer>> accessibleConstantDependents;
+    private final Map<Integer, IntSet> privateConstantDependents;
+    private final Map<Integer, IntSet> accessibleConstantDependents;
 
     ConstantToDependentsMapping(Set<String> visitedClasses,
                                         List<String> dependentClasses,
-                                        Map<Integer, Set<Integer>> privateConstantDependents,
-                                        Map<Integer, Set<Integer>> accessibleConstantDependents) {
+                                        Map<Integer, IntSet> privateConstantDependents,
+                                        Map<Integer, IntSet> accessibleConstantDependents) {
         this.visitedClasses = visitedClasses;
         this.dependentClasses = dependentClasses;
         this.privateConstantDependents = privateConstantDependents;
@@ -69,11 +71,11 @@
         return visitedClasses;
     }
 
-    public Map<Integer, Set<Integer>> getAccessibleConstantDependents() {
+    public Map<Integer, IntSet> getAccessibleConstantDependents() {
         return accessibleConstantDependents;
     }
 
-    public Map<Integer, Set<Integer>> getPrivateConstantDependents() {
+    public Map<Integer, IntSet> getPrivateConstantDependents() {
         return privateConstantDependents;
     }
 
@@ -85,9 +87,9 @@
         return findDependentsFor(accessibleConstantDependents, constantOriginHash);
     }
 
-    private Set<String> findDependentsFor(Map<Integer, Set<Integer>> collection, int constantOriginHash) {
+    private Set<String> findDependentsFor(Map<Integer, IntSet> collection, int constantOriginHash) {
         if (collection.containsKey(constantOriginHash)) {
-            Set<Integer> classIndexes = collection.get(constantOriginHash);
+            IntSet classIndexes = collection.get(constantOriginHash);
             Set<String> dependents = new ObjectOpenHashSet<>(classIndexes.size());
             classIndexes.forEach(index -> dependents.add(dependentClasses.get(index)));
             return dependents;
@@ -105,7 +107,7 @@ public static ConstantToDependentsMappingBuilder builder() {
 
     public static final class Serializer extends AbstractSerializer<ConstantToDependentsMapping> {
 
-        private final Int2ObjectMapSerializer<Set<Integer>> mapSerializer;
+        private final Int2ObjectMapSerializer<IntSet> mapSerializer;
         private final ListSerializer<String> classNamesSerializer;
 
         public Serializer(StringInterner interner) {
@@ -117,8 +119,8 @@ public Serializer(StringInterner interner) {
         @Override
         public ConstantToDependentsMapping read(Decoder decoder) throws Exception {
             List<String> classNames = classNamesSerializer.read(decoder);
-            Map<Integer, Set<Integer>> privateDependentsIndexes = mapSerializer.read(decoder);
-            Map<Integer, Set<Integer>> publicDependentsIndexes = mapSerializer.read(decoder);
+            Map<Integer, IntSet> privateDependentsIndexes = mapSerializer.read(decoder);
+            Map<Integer, IntSet> publicDependentsIndexes = mapSerializer.read(decoder);
             return new ConstantToDependentsMapping(Collections.emptySet(), classNames, privateDependentsIndexes, publicDependentsIndexes);
         }
 
@@ -130,15 +132,15 @@ public void write(Encoder encoder, ConstantToDependentsMapping value) throws Exc
         }
     }
 
-    private static final class JavaSetToIntSetSerializer implements org.gradle.internal.serialize.Serializer<Set<Integer>> {
+    private static final class JavaSetToIntSetSerializer implements org.gradle.internal.serialize.Serializer<IntSet> {
         @Override
-        public Set<Integer> read(Decoder decoder) throws Exception {
+        public IntSet read(Decoder decoder) throws Exception {
             return IntSetSerializer.INSTANCE.read(decoder);
         }
 
         @Override
-        public void write(Encoder encoder, Set<Integer> value) throws Exception {
-            IntSetSerializer.INSTANCE.write(encoder, new IntOpenHashSet(value));
+        public void write(Encoder encoder, IntSet value) throws Exception {
+            IntSetSerializer.INSTANCE.write(encoder, value);
         }
     }
 
diff --git a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/compilerapi/constants/ConstantToDependentsMappingBuilder.java b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/compilerapi/constants/ConstantToDependentsMappingBuilder.java
index f65723d23bb..87c55b1f322 100644
--- a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/compilerapi/constants/ConstantToDependentsMappingBuilder.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/compilerapi/constants/ConstantToDependentsMappingBuilder.java
@@ -16,6 +16,9 @@
 
 package org.gradle.api.internal.tasks.compile.incremental.compilerapi.constants;
 
+import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
+import it.unimi.dsi.fastutil.ints.IntSet;
+
 import java.io.Serializable;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -33,8 +36,8 @@
 
     private final Set<String> visitedClasses;
     private final List<String> dependents;
-    private final Map<Integer, Set<Integer>> privateDependentsIndexes;
-    private final Map<Integer, Set<Integer>> accessibleDependentsIndexes;
+    private final Map<Integer, IntSet> privateDependentsIndexes;
+    private final Map<Integer, IntSet> accessibleDependentsIndexes;
     private final Map<String, Integer> classNameToIndex;
 
     ConstantToDependentsMappingBuilder() {
@@ -56,8 +59,8 @@ public ConstantToDependentsMappingBuilder addPrivateDependents(int constantOrigi
     }
 
     public ConstantToDependentsMappingBuilder addPrivateDependent(int constantOriginHash, String dependent) {
-        Set<Integer> accessibleDependents = accessibleDependentsIndexes.getOrDefault(constantOriginHash, Collections.emptySet());
-        Set<Integer> privateDependents = privateDependentsIndexes.computeIfAbsent(constantOriginHash, k -> new HashSet<>());
+        IntSet accessibleDependents = accessibleDependentsIndexes.getOrDefault(constantOriginHash, IntSet.of());
+        IntSet privateDependents = privateDependentsIndexes.computeIfAbsent(constantOriginHash, k -> new IntOpenHashSet());
         int dependentIndex = classNameToIndex.getOrDefault(dependent, -1);
         if (dependentIndex < 0 || !accessibleDependents.contains(dependentIndex)) {
             addDependent(privateDependents, dependent);
@@ -66,8 +69,8 @@ public ConstantToDependentsMappingBuilder addPrivateDependent(int constantOrigin
     }
 
     public ConstantToDependentsMappingBuilder addAccessibleDependent(int constantOriginHash, String dependent) {
-        Set<Integer> accessibleDependents = accessibleDependentsIndexes.computeIfAbsent(constantOriginHash, k -> new HashSet<>());
-        Set<Integer> privateDependents = privateDependentsIndexes.getOrDefault(constantOriginHash, Collections.emptySet());
+        IntSet accessibleDependents = accessibleDependentsIndexes.computeIfAbsent(constantOriginHash, k -> new IntOpenHashSet());
+        IntSet privateDependents = privateDependentsIndexes.getOrDefault(constantOriginHash, IntSet.of());
         int dependentIndex = addDependent(accessibleDependents, dependent);
         if (!privateDependents.isEmpty()) {
             privateDependents.remove(dependentIndex);
@@ -80,7 +83,7 @@ public ConstantToDependentsMappingBuilder addVisitedClass(String visitedClass) {
         return this;
     }
 
-    private int addDependent(Set<Integer> dependentIndexes, String dependent) {
+    private int addDependent(IntSet dependentIndexes, String dependent) {
         int dependentIndex = classNameToIndex.computeIfAbsent(dependent, k -> {
             addVisitedClass(dependent);
             dependents.add(dependent);
```

Reply:
Does GroovyJavaJointIncrementalCompilationIntegrationTest pass if you use IntSet? I did not use IntSet in builder, since that test failed with NoClassFoundException, but you probably know asap how to fix this. Do you have any hint how to deal with it?

Reply:
Hmm, from git blame it seems that incremental compiler should not be triggered in that case. 
```
Currently, incremental Groovy compilation only works for files compiled by the Groovy compiler. Java files are compiled by the Java compiler. This PR disables incremental compiler if Java files found because this might cause incorrectness.
```
Is that the case just for Groovy compiler, or also for Java?